### ISSUE TITLE
PDF output size

### ISSUE THREAD

**Issue by neodynamic** [mono/SkiaSharp#848]
Running this sample https://github.com/mono/SkiaSharp/blob/master/samples/Gallery/Shared/Samples/CreatePdfSample.cs which creates a simple two pages PDF file, the created file (under Windows) is about 510KB

Is there any compression setting to get the output PDF file size lighter? 510KB for a two pages PDF with a simple text seems to be somehow heavy... Any hints?

**Comment by charlesroddie** [comment]
@neodynamic Can you see what [Skia](https://github.com/google/skia/blob/master/site/user/sample/pdf.md) produces?

**Comment by neodynamic** [comment]
@charlesroddie no, and you?

**Comment by mattleibow** [comment]
You could have a look at setting the quality down: https://docs.microsoft.com/en-us/dotnet/api/skiasharp.skdocumentpdfmetadata

**Comment by neodynamic** [comment]
We've been reviewing this matter and we can conclude that the concerns about pdf output file size cannot be improved because the following... it seems that Skia (native lib) [PDF backend design](https://skia.org/dev/design/pdftheory#Fonts) will embed any font file needed to render the text at the target device. That page states the following:

> We can't assume that an arbitrary font will be available at PDF view time, so we embed all fonts in accordance with modern PDF guidelines.

The sample here [https://github.com/mono/SkiaSharp/blob/master/samples/Gallery/Shared/Samples/CreatePdfSample.cs](https://github.com/mono/SkiaSharp/blob/master/samples/Gallery/Shared/Samples/CreatePdfSample.cs) will use the default font in the system, which under Windows, it's likely to be `Segoe UI` which TTF file size is about 900KB
The output PDF file for that simple test where a single text is drawn is about 510KB. That big size for such a simple pdf is because `Segoe UI` font file is embedded in the file by Skia design. We've made another tests by drawing Chinese text using the `Yu Gothic` font which file size is about 13MB! and the output pdf file is about 8MB! which confirms that the size is because the font files being embedded. Linked fonts seems not to be supported which could make the pdf output file size smaller. 

If no one here has more comments on this matter, then @mattleibow you can close this issue.

**Comment by Gillibald** [comment]
You could use HarfBuzz's subsetting to reduce the font's size. Then load that font to produce the PDF. Sadly this isn't supported by HarfBuzzSharp. Yet....

**Comment by neodynamic** [comment]
Yes, that could be the only way to reduce pdf output file size...

**Comment by mattleibow** [comment]
Looking at the skia code, it seems there is 2 subsetters built in. But, this is disabled because we are not building with either icu or harfbuzz/sfntly.

However, there is a hook that makes subsetting work, but it is not a "public API". But, since it is fairly simple, we might be able to do something. The API hasn't changed much, so it might just be safe to do something. 

I'll have a look at what we can do. Can't promise anything as I haven't had a look at exactly how the PDF is constructed, but it seems to only write the fonts when the PDF is closed, so we could potentially add a argument there, or in the metadata in the constructor. They actually have an enum there that allows you to pick either harfbuzz or sfntly. Seems to be not too hard to add one for us, and then we can use _any_ font subsetter.

**Comment by mattleibow** [comment]
Started a thing on the skia bugtracker. I want to do this right: https://bugs.chromium.org/p/skia/issues/detail?id=10491

**Comment by Alexbits** [comment]
> You could have a look at setting the quality down: https://docs.microsoft.com/en-us/dotnet/api/skiasharp.skdocumentpdfmetadata

I've tried to set lower EncodingQuality and RasterDpi and they have no impact on the output file at all. It outputs the same file size and quality. Latest SkiaSharp on Windows 10.

**Comment by reinux** [comment]
Any progress on this? Japanese/Chinese fonts are 5MB+, so this becomes nigh unusable.

**Comment by wstaelens** [comment]
any updates regarding the file size/fonts?

**Comment by reinux** [comment]
Should I assume this has been abandoned and rebuild my project using another PDF library?

**Comment by johmarjac** [comment]
> Should I assume this has been abandoned and rebuild my project using another PDF library?

In case you go for a different library, don't use QuestPDF as it uses SkiaSharp under the hood and suffers from the same big file sizes.

**Comment by reinux** [comment]
Thanks for the tip. I don't understand how something like this wouldn't be recognized as a fatal issue.

> If no one here has more comments on this matter, then @mattleibow you can close this issue.

Like, what.

**Comment by johmarjac** [comment]
> Thanks for the tip. I don't understand how something like this wouldn't be recognized as a fatal issue.
> 
> > If no one here has more comments on this matter, then @mattleibow you can close this issue.
> 
> Like, what.

Depends on the use case really. If you only generate a single pdf, no one cares for a 2 MB PDF on their PC. But I needed it for a production series for part protocol where I have a part every 2 seconds so every 2 seconds I need to save a pdf to network share to archive the part measure results. Then a 2 MB file every 2 seconds costs a hell lot of storage and that's just not going to work

**Comment by wstaelens** [comment]
indeed, thousands of small files are processed. 1000 * 2MB (while it is normally like 159KB - 236KB really means a big difference, in network traffic, processing time, diskspace etc...)

it is related to fonts, but should be investigated by skia...

**Comment by KillyMXI** [comment]
Is there a way to force Skia to render all text as paths?

**Comment by KillyMXI** [comment]
I'm trying to run [pdf checker](https://www.datalogics.com/products/pdf-tools/pdf-checker/) on a file generated with SkiaSharp, with no strings in it.
The only non-empty notice in the report:

```
Cleanup Results
    Errors:
        None
    Information:
        Contains conservatively compressed streams:
            Uncompressed: (141 instances)
    Checks Completed:
        suboptimal-compression
```

Optimizing it with 3rd party tools allowed to go from ~500kb to 100kb.

Looks like there is something besides embedded fonts that could be optimized in Skia.

My sample is generated with Svg.Skia and the source only consists of vector lines. I'm not sure what is so inefficient there.

Trying to mess with SKDocumentPdfMetadata actually results in bigger file size. I would expect it to be a no-op, but if I supply any RasterDpi value or non-default EncodingQuality value, the file size jumps up another ~400kb. This doesn't make sense.

**Comment by domagojmedo** [comment]
Any updates for this? 
@mattleibow

**Comment by Greybird** [comment]
> I ended up using a two-pass approach by building the font subsets before rendering and then passing those in to SkiaSharp.
@jeffska : would you have a gist or some place where we could take a look at what you put in place to build the font subsets externally ?

**Comment by TimLee88** [comment]
So, How is the progress?

**Comment by wstaelens** [comment]
Wondering the same @TimLee88

**Comment by wstaelens** [comment]
images in PDFs don't seem to support 1bpp which increases also the pdf size, correct?

**Comment by giz303** [comment]
Hi, has anything happened here? We are looking for a solution.
We had been using https://github.com/Sicos1977/ChromeHtmlToPdf to convert from SVG to PDF and moved to SkiaSharp to get rid of the Google chrome processes.

But now the PDF files which had been between 20 and 40 KB are now over 500 KB big.
Since we convert a lot of files in production, and need to send these PDF files over ethernet to terminals, we would like the file sizes to be lower again.
All used fonts are available on the terminals and there is no need to embedded them in the PDF file.

So, is there someone workling on this issue or will this not be implemented at all?

**Comment by jeffska** [comment]
> Hi, has anything happened here? We are looking for a solution. We had been using https://github.com/Sicos1977/ChromeHtmlToPdf to convert from SVG to PDF and moved to SkiaSharp to get rid of the Google chrome processes.
> 
> But now the PDF files which had been between 20 and 40 KB are now over 500 KB big. Since we convert a lot of files in production, and need to send these PDF files over ethernet to terminals, we would like the file sizes to be lower again. All used fonts are available on the terminals and there is no need to embed them in the PDF file.

Have you inspected the PDF to make sure the SVG isn't just being rasterized?

**Comment by wstaelens** [comment]
we've seen an increase because 1bpp images are not supported. Resulting in larger pdf's (every 1bpp image is converted to 24bpp). The 1bpp pdf's happen when multifunctional devices make scans..

Would like to see some support also for 1bpp... as this makes pdf's much much bigger. (especially when 1bpp glyph bitmaps are used)

**Comment by flensrocker** [comment]
I don't care so much about performance or lib size and I already use HarfBuzzSharp for measuring text widths.
Is there any way to (optionally) enable font subsetting with HarfBuzz?

**Comment by cimpok** [comment]
It was yesterday I chose SkiaSharp for my back-end api server to create small flowcharts-like pages (svg for in-browser view, and pdf for printing of the same thing)   My reason to choose SkiaSharp was that it can do both.

Today it took me four hours to track down that it is not me doing anything so wrong that a one-page pdf with a single rectangle and a label in it is half a megabyte large, the same is below 1KB in svg.  (I need hundreds of these on my server, and my users would download them dozens per session )   I was hoping to be in the ballpark of 10-50KB per file the maximum.

...and then I just found out here that in the FIVE years since 2019 when this issue was first mentioned here, nothing happened, except that the issue was closed. I guess it would be not much more complicated than putting in an extra if statemen in the code.
I gave up on SkiSharp, removed the NuGet package from my project.   Trying PdfSharp and  SVG.NET now..
...Btw, is open source dying?

**Comment by reinux** [comment]
> ...Btw, is open source dying?

I'd say that in general, ongoing economic consolidation and things like covid have probably made anything free or Free or "free" less sustainable.

That said, this is a reasonably reasonably active repo, so I can only conclude that they just don't care about pdf output.

**Comment by wstaelens** [comment]
one of the problems is that these are not supported: https://github.com/mono/SkiaSharp/issues/848#issuecomment-1926427146


Note: Microsoft is putting again effort on XPS after ±8 years. Might be an option to switch again to XPS files 🙌

**Comment by 18779758522** [comment]
Since Skia cannot enable font subset, can we choose not to embed font files when generating PDFs? Or delete the embedded fonts after generating the PDF. Conventional fonts exist in Windows systems, and deleting embedded fonts has no impact on PDFs

**Comment by 18779758522** [comment]
> > Skia 无法启用字体子集，那么是否可以在生成 PDF 时选择不嵌入字体文件？或者在生成 PDF 后删除嵌入的字体。Windows 系统中存在常规字体，删除嵌入字体对 PDF 没有影响
> 
> 不嵌入字体_会对_打印产生巨大的负面影响，因为打印依赖于嵌入的字体信息。
> 
> 需要支持 1bpp 图像来减小尺寸 :-)

/

> > Since Skia cannot enable font subset, can we choose not to embed font files when generating PDFs? Or delete the embedded fonts after generating the PDF. Conventional fonts exist in Windows systems, and deleting embedded fonts has no impact on PDFs
> 
> not embedding fonts _can_ have a huge negative impact on/when printing, which relies on the font information embedded.
> 
> supporting 1bpp images would be needed to reduce size :-)

Have you tried deleting embedded fonts?

**Comment by 18779758522** [comment]
> > Since Skia cannot enable font subset, can we choose not to embed font files when generating PDFs? Or delete the embedded fonts after generating the PDF. Conventional fonts exist in Windows systems, and deleting embedded fonts has no impact on PDFs
> 
> not embedding fonts _can_ have a huge negative impact on/when printing, which relies on the font information embedded.
> 
> supporting 1bpp images would be needed to reduce size :-)

Have you tried deleting embedded fonts?

**Comment by wstaelens** [comment]
> > > Since Skia cannot enable font subset, can we choose not to embed font files when generating PDFs? Or delete the embedded fonts after generating the PDF. Conventional fonts exist in Windows systems, and deleting embedded fonts has no impact on PDFs
> > 
> > 
> > not embedding fonts _can_ have a huge negative impact on/when printing, which relies on the font information embedded.
> > supporting 1bpp images would be needed to reduce size :-)
> 
> Have you tried deleting embedded fonts?

another optimization should be to check if the same font/font tables are added per page, if they can be set centrally to reduce the resources per page.

**Comment by jbartlau** [comment]
> [@mattleibow](https://github.com/mattleibow) I was able to build the Windows libSkiaSharp using Skia's support for Harfbuzz subsetting. It seems to work fine.

@jeffska did you ever think of opening a PR for this change? Would this be welcome, @mattleibow ? This is such a great project, however the PDF file size issue keeps coming back all the time. I'd rather not have to fork and lose the benefits of main line support. However I can offer to create a PR and do thorough testing if there's a chance for getting a merge.
